File: common/src/main/java/munoon/bank/common/validation/money/ValidMoneyCount.java
package munoon.bank.common.validation.money;


import javax.validation.Constraint;
import javax.validation.Payload;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = MoneyCountValidator.class)
@Target({ElementType.FIELD, ElementType.PARAMETER})
public @interface ValidMoneyCount {
    String message() default "Сумма указана неверно";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: transactional-service/src/main/kotlin/munoon/bank/service/transactional/util/validation/ValidMoneyCount.kt
package munoon.bank.service.transactional.util.validation

import javax.validation.Constraint
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext
import javax.validation.Payload
import kotlin.reflect.KClass

@MustBeDocumented
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [MoneyCountValidator::class])
@Target(AnnotationTarget.FIELD, AnnotationTarget.VALUE_PARAMETER)
annotation class ValidMoneyCount(
        val message: String = "Сумма указана неверно",
        val groups: Array<KClass<*>> = [],
        val payload: Array<KClass<out Payload>> = []
)

class MoneyCountValidator : ConstraintValidator<ValidMoneyCount, Double> {
    override fun isValid(value: Double?, context: ConstraintValidatorContext): Boolean {
        val string = value?.toString() ?: return true
        val index = string.indexOf(".")
        return if (index == -1) true
               else string.substring(index).length <= 2
    }
}